File: scala/src/main/scala/folio/codinginterview/presentation/PresentationPreparation.scala
package folio.codinginterview.presentation

import folio.codinginterview.domain.UserId
import folio.codinginterview.presentation.PresentationException.BadRequestException
import scala.concurrent.Future
import scala.util.control.NonFatal

trait PresentationPreparation {
  protected def parseUserId(s: String): Future[UserId] =
    try Future.successful(UserId(s))
    catch { case NonFatal(e) => Future.failed(BadRequestException(e.getMessage)) }

  protected def parseAmount(s: String): Future[BigDecimal] =
    try Future.successful(BigDecimal(s))
    catch { case NonFatal(_) => Future.failed(BadRequestException(s"invalid amount: $s")) }
}

File: php/src/CodingInterview/Presentation/PresentationPreparation.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Presentation;

use Folio\CodingInterview\Domain\BigDecimal;
use Folio\CodingInterview\Domain\UserId;

trait PresentationPreparation
{
    protected function parseUserId(string $s): UserId
    {
        try {
            return new UserId($s);
        } catch (\Throwable $e) {
            throw new BadRequestException($e->getMessage());
        }
    }

    protected function parseAmount(string $s): BigDecimal
    {
        try {
            return new BigDecimal($s);
        } catch (\Throwable $e) {
            throw new BadRequestException("invalid amount: {$s}");
        }
    }
}
